types/funds.go

```diff
@@ -42,11 +42,40 @@ func (h Funds) Add(a ...Funds) Funds {
 func Sum(a ...Funds) Funds {
 	sum := Funds{}
 
-	for _, holdings := range a {
-		for asset, amount := range holdings {
+	for _, funds := range a {
+		for asset, amount := range funds {
+			if sum[asset] == nil {
+				sum[asset] = big.NewInt(0)
+			}
+
 			sum[asset] = sum[asset].Add(sum[asset], amount)
 		}
 	}
 
 	return sum
 }
+
+// Equal returns true if reciever `f` and input `g` are identical in value.
+//
+// Note that a zero-balance equals a non-balance: {[0x0a,0x00],[0x0b,0x01]} == {[0x0b,0x01]}
+func (f Funds) Equal(g Funds) bool {
+	return f.canAfford(g) && g.canAfford(f)
+}
```

Comment:
1. We discussed whether we want to consider zero-balances with non-balances. I think we know about some situations where we wouldn't that kind of behaviour -- for example where the respective hashes of each side in the "equality"  are important.  That said, I'm fine with this implementation for now. It's well documented and we can amend or extend it later.
2. The implementation is pretty elegant (calling into `canAfford`). But I guess it's also a bit surprising, and probably also a bit less efficient than a more naive approach? Interested to know your thoughts.

Reply:
Re. 1, as long as there's a mapping from `Funds` to `Outcome` such that `f.Equals(g)` implies `hash(f.AsOutcome) == hash(g.AsOutcome)`, we're fine.

Considering this, defining `f.Equals(g)` as `hash(f.AsOutcome) == hash(g.AsOutcome)` may make sense, but is probably _less_ efficient, as it relies on cryptography.

We are currently not concerned with efficiency, so IMO it could go either way. I like the idea of avoiding cryptography if there's a safe alternative implementation, but we can always implement the safe, efficient alternative at a later point, and so I would lean towards the _obviously correct_ implementation.

Reply:
I think `Funds` is primarily there to represent on-chain `holdings` rather than on chain `Outcomes`.  As long as we keep that separation clear, we shouldn't run into trouble with the current notion of `Funds.Equal`.

If we want to convert from `Funds` to `Outcomes`, then I think we need the stricter / more opinionated notion that @andrewgordstewart mentions.

Reply:
Seeing now that a response I had written was not submitted and no longer exists... BUT, @geoknee has cleanly articulated the separation between `Funds` and `Outcome`s.

The other point was about **2)**. This `Equal` method naturally fell out of the more naive approach. Golang doesn't have friendly access to map keys, unions of collections, etc, so most any implementation will involve looping over the keys from both sides.  After doing that, it was natural to refactor the symmetric operation into its own function, which happens to be the semantically meaningful `canAfford`.